File: questions/3.LongestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = function (s) {
  const set = new Set()
  let left = 0
  let result = 0

  for (let right = 0; right < s.length; right++) {
    // e.g. `pww`
    //  forward the left to `w` index.
    while (set.has(s[right])) {
      set.delete(s[left])
      left++
    }
    set.add(s[right])
    result = Math.max(result, right - left + 1)
  }

  return result
}
console.log(lengthOfLongestSubstring('pwwkew'))

File: leetcode/3.LongestSubstringWithoutRepeatingCharacters.go
package leetcode

import "math"

func lengthOfLongestSubstring(s string) int {
	set := make(map[string]struct{})
	left := 0
	result := 0

	for right := range s {
		length := len(set)
		for i := 0; i < length; i++ {
			_, ok := set[string(s[right])]
			if ok {
				delete(set, string(s[left]))
				left++
			}
		}

		set[string(s[right])] = struct{}{}
		result = int(math.Max(float64(result), float64(right-left+1)))
	}

	return result
}
